Unity-Technologies/usd-unity-sdk
# USDU-46 create sparse timesamples

## Purpose of this PR

**Ticket/Jira #:** USDU-46



Fix redundant timesamples created every frame for objects that are not animated, or whose animation is not changing.

## Testing

**Functional Testing status:**


Added a unit test for recording a simple hierarchy to USD from the timeline and verifying that only the animated objects contain timesamples in the USD file.

Manual Test 1:
1. Create a cube and a sphere parented under it (or test with a different file that has at least one unanimated object)
2. Create a timeline and add the sphere to it
3. Animate the sphere but not the cube
4. Add a USD recorder to the timeline
5. Set the USD recorder export root to be the cube
6. Click play to record
7. Stop playing
8. Inspect recorded file, check that there are no timesamples for the cube

Manual Test 2:
1. Run Manual Test 1 with the HumanFemale walking asset
2. Import into both Unity and Maya -> check that it imports upright
Note: there are some issues with the animation, though I think they were present before the fix as well.

[recording_person_before_and_after_fix.zip](https://github.com/Unity-Technologies/usd-unity-sdk/files/5461249/recording_person_before_and_after_fix.zip)

Manual Test 3:
1. Run the Export Mesh sample
2. Check that it gives the correct result

**Performance Testing status:**


Profiled export of HumanFemale walking asset with and without change. Gave similar results in profiler.

## Overall Product Risks


**Complexity:**

Low

**Halo Effect:**

Medium - changes the way attribute values are written (through the UsdUtilsSparseAttrValueWriter as opposed to written directly with attr.SetValue()).

## Additional information

**Note to reviewers:**



**Reminder:**
